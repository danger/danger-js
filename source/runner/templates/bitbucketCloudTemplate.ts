import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../dsl/Violation"
import { compliment } from "../DangerUtils"

// BitBucket Cloud supports these emojis 🎉
const noEntryEmoji = "❌"
const warningEmoji = "⚠️"
const messageEmoji = "✨"
const signatureEmoji = "🚫"
const successEmoji = "🎉"

export const dangerSignature = (results: DangerResults) => {
  let meta = results.meta || { runtimeName: "dangerJS", runtimeHref: "https://danger.systems/js" }
  return `Generated by ${signatureEmoji} [${meta.runtimeName}](${meta.runtimeHref})`
}

export const messageForResultWithIssues = `${warningEmoji}  Danger found some issues. Don't worry, everything is fixable.`

export const dangerIDToString = (id: string) => `danger-id-${id};`
export const fileLineToString = (file: string, line: number) => `  File: ${file};Line: ${line};`
/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
export const dangerSignaturePostfix = (results: DangerResults, commitID?: string) => {
  let signature = dangerSignature(results)
  if (commitID !== undefined) {
    signature = `${signature} against ${commitID}`
  }
  return `

  ${signature}
  `
}

function buildMarkdownTable(header: string, defaultEmoji: string, violations: Violation[]): string {
  if (violations.length === 0 || violations.every(violation => !violation.message)) {
    return ""
  }
  return `

  |      ${violations.length} ${header} |
  | --- |
${violations.map(v => `  | ${v.icon || defaultEmoji} - ${v.message} |`).join("\n")}

  `
}

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {string} commitID The hash that represents the latest commit
 * @param {DangerResults} results Data to work with
 * @returns {string} HTML
 */
export function template(dangerID: string, results: DangerResults, commitID?: string): string {
  let summaryMessage: string
  if (!results.fails.length && !results.warnings.length) {
    summaryMessage = `${successEmoji}  All green. ${compliment()}`
  } else {
    summaryMessage = messageForResultWithIssues
  }
  return `
  ${summaryMessage}

  ${buildMarkdownTable("Fails", noEntryEmoji, results.fails)}
  ${buildMarkdownTable("Warnings", warningEmoji, results.warnings)}
  ${buildMarkdownTable("Messages", messageEmoji, results.messages)}
  
  ${results.markdowns.map(v => v.message).join("\n\n")}
  
  ${dangerSignaturePostfix(results, commitID)}
  
  [](http://${dangerIDToString(dangerID)})
  `
}

export function inlineTemplate(dangerID: string, results: DangerResults, file: string, line: number): string {
  const printViolation = (defaultEmoji: string) => (violation: Violation) => {
    return `- ${violation.icon || defaultEmoji} ${violation.message}`
  }

  return `
[//]: # (${dangerIDToString(dangerID)})
[//]: # (${fileLineToString(file, line)})
${results.fails.map(printViolation(noEntryEmoji)).join("\n")}
${results.warnings.map(printViolation(warningEmoji)).join("\n")}
${results.messages.map(printViolation(messageEmoji)).join("\n")}
${results.markdowns.map(v => v.message).join("\n\n")}
  `
}
