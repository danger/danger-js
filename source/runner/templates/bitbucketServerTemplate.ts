import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../dsl/Violation"
import {
  dangerIDToString,
  messageForResultWithIssues,
  dangerSignature,
  noEntryEmoji,
  warningEmoji,
  messageEmoji,
} from "./bitbucketTemplateCommon"

/**
 * Converts a set of violations into a Markdown section
 *
 * @param {string} name User facing title of section
 * @param {string} emoji Emoji name to show
 * @param {Violation[]} violations for section
 * @returns {string} Markdown
 */
function resultsSection(name: string, emoji: string, violations: Violation[]): string {
  if (violations.length === 0 || violations.every(violation => !violation.message)) {
    return ""
  }
  return (
    `\n` +
    `| ${emoji} | ${name}\n` +
    `| --- | --- |\n` +
    `\n` +
    violations
      .map(v => {
        return (
          "> " +
          v.message
            .replace(/<\/?code>/g, "`")
            .split("\n")
            .join("\n> ")
        )
      })
      .join("\n\n") +
    `\n`
  )
}

/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
export const dangerSignaturePostfix = (results: DangerResults, commitID?: string) => {
  let signature = dangerSignature(results)
  if (commitID !== undefined) {
    signature = `${signature} against ${commitID}`
  }
  return `
|    |
|---:|
| _${signature}_ |
`
}

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {DangerResults} results Data to work with
 * @param {string} commitID The hash that represents the latest commit
 * @returns {string} HTML
 */
export function template(dangerID: string, results: DangerResults, commitID?: string): string {
  return `
${messageForResultWithIssues}

${resultsSection("Fails", noEntryEmoji, results.fails)}
${resultsSection("Warnings", warningEmoji, results.warnings)}
${resultsSection("Messages", messageEmoji, results.messages)}

${results.markdowns.map(v => v.message).join("\n\n")}

${dangerSignaturePostfix(results, commitID)}

[](http://${dangerIDToString(dangerID)})
`
}
