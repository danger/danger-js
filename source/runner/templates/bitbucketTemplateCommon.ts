import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../dsl/Violation"

export const noEntryEmoji = "\u274C"
export const warningEmoji = "⚠️"
export const messageEmoji = "\u2728"
export const signatureEmoji = "\u1F6AB"

export const dangerIDToString = (id: string) => `danger-id-${id};`
export const fileLineToString = (file: string, line: number) => `  File: ${file};Line: ${line};`

/**
 * Comment to add when updating the PR status when issues are found
 */
export const messageForResultWithIssues = `${warningEmoji}  Danger found some issues. Don't worry, everything is fixable.`

export function inlineTemplate(dangerID: string, results: DangerResults, file: string, line: number): string {
  const printViolation = (defaultEmoji: string) => (violation: Violation) => {
    return `- ${violation.icon || defaultEmoji} ${violation.message}`
  }

  return `
[//]: # (${dangerIDToString(dangerID)})
[//]: # (${fileLineToString(file, line)})
${results.fails.map(printViolation(noEntryEmoji)).join("\n")}
${results.warnings.map(printViolation(warningEmoji)).join("\n")}
${results.messages.map(printViolation(messageEmoji)).join("\n")}
${results.markdowns.map(v => v.message).join("\n\n")}
  `
}

export const dangerSignature = (results: DangerResults, icon: string) => {
  let meta = results.meta || { runtimeName: "dangerJS", runtimeHref: "https://danger.systems/js" }
  return `Generated by ${icon} [${meta.runtimeName}](${meta.runtimeHref})`
}
