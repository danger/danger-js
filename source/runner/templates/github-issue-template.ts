import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../platforms/messaging/violation"
import * as v from "voca"

/**
 * Converts a set of violations into a HTML table
 *
 * @param {string} name User facing title of table
 * @param {string} emoji Emoji name to show next to each item
 * @param {Violation[]} violations for table
 * @returns {string} HTML
 */
function table(name: string, emoji: string, violations: Array<Violation>): string {
  if (violations.length === 0) { return "" }
  return `
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">${name}</th>
    </tr>
  </thead>
  <tbody>${violations.map((v: Violation) => { return `<tr>
      <td>:${emoji}:</td>
      <td>${v.message}</td>
    </tr>
  ` }).join("\n")}</tbody>
</table>
`
}

function getSummary(label: string, violations: Array<Violation>): string {
  return violations.map(x => v.truncate(x.message, 20))
    .reduce((acc, value, idx) => `${acc} ${value}${idx === violations.length - 1 ? "" : ","}`, `${violations.length} ${label}: `)
}

function buildSummaryMessage(results: DangerResults): string {
  const {fails, warnings, messages, markdowns } = results
  const summary =
`  ${getSummary("failure", fails)}
  ${getSummary("warning", warnings)}
  ${messages.length > 0 ? `${messages.length} messages` : ""}
  ${markdowns.length > 0 ? `${markdowns.length} markdown notices` : ""}`
  return summary
}

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {DangerResults} results Data to work with
 * @returns {string} HTML
 */
export function template(results: DangerResults): string {
  return `
<!--
${buildSummaryMessage(results)}
-->
${table("Fails", "no_entry_sign", results.fails)}
${table("Warnings", "warning", results.warnings)}
${table("Messages", "book", results.messages)}
${results.markdowns.map(markdown => markdown.message).join("\n\n")}
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>
`
}
