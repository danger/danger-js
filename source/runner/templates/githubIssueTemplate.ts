import { DangerResults } from "../../dsl/DangerResults"
import { Violation, isInline } from "../../dsl/Violation"

/**
 * Converts a set of violations into a HTML table
 *
 * @param {string} name User facing title of table
 * @param {string} defaultEmoji Emoji name to show next to each item
 * @param {Violation[]} violations for table
 * @returns {string} HTML
 */
function table(name: string, defaultEmoji: string, violations: Violation[]): string {
  if (noViolationsOrAllOfThemEmpty(violations)) {
    return ""
  }
  return `
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">${name}</th>
    </tr>
  </thead>
  <tbody>${violations.map(violation => htmlForValidation(defaultEmoji, violation)).join("\n")}</tbody>
</table>
`
}

function htmlForValidation(defaultEmoji: string, violation: Violation) {
  let message = isInline(violation)
    ? `**${violation.file!}#L${violation.line!}** - ${violation.message}`
    : violation.message

  if (containsMarkdown(message)) {
    message = `\n\n  ${message}\n  `
  }

  const emojiString = `:${defaultEmoji}:`;

  return `<tr>
      <td>${violation.icon || emojiString}</td>
      <td>${message}</td>
    </tr>
  `
}

function containsMarkdown(message: string): boolean {
  return message.match(/[`*_~\[]+/g) ? true : false
}

function noViolationsOrAllOfThemEmpty(violations: Violation[]) {
  return violations.length === 0 || violations.every(violation => !violation.message)
}

const truncate = (msg: string, count: number) => {
  if (!msg) {
    return msg
  }
  if (msg.length < count) {
    return msg
  } else {
    return msg.substr(0, count - 3) + "..."
  }
}

function getSummary(label: string, violations: Violation[]): string {
  return violations
    .map(x => truncate(x.message, 20))
    .reduce(
      (acc, value, idx) => `${acc} ${value}${idx === violations.length - 1 ? "" : ","}`,
      `${violations.length} ${label}: `
    )
}

function buildSummaryMessage(dangerID: string, results: DangerResults): string {
  const { fails, warnings, messages, markdowns } = results
  const summary = `  ${getSummary("failure", fails)}
  ${getSummary("warning", warnings)}
  ${messages.length > 0 ? `${messages.length} messages` : ""}
  ${markdowns.length > 0 ? `${markdowns.length} markdown notices` : ""}
  ${dangerIDToString(dangerID)}`
  return summary
}

export const dangerIDToString = (id: string) => `DangerID: danger-id-${id};`
export const fileLineToString = (file: string, line: number) => `  File: ${file};
  Line: ${line};`

export const dangerSignature = (results: DangerResults) => {
  let meta = results.meta || { runtimeName: "dangerJS", runtimeHref: "https://danger.systems/js" }

  return `Generated by :no_entry_sign: <a href="${meta.runtimeHref}">${meta.runtimeName}</a>`
}

/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
export const dangerSignaturePostfix = (results: DangerResults, commitID?: string) => {
  let signature = dangerSignature(results)
  if (commitID !== undefined) {
    signature = `${signature} against ${commitID}`
  }
  return signature
}

/**
 * Comment to add when updating the PR status when issues are found
 */
export const messageForResultWithIssues = `Found some issues. Don't worry, everything is fixable.`

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {DangerResults} results Data to work with
 * @param {string} commitID The hash that represents the latest commit
 * @returns {string} HTML
 */
export function template(dangerID: string, results: DangerResults, commitID?: string): string {
  return `
<!--
${buildSummaryMessage(dangerID, results)}
-->
${table("Fails", "no_entry_sign", results.fails)}
${table("Warnings", "warning", results.warnings)}
${table("Messages", "book", results.messages)}
${results.markdowns.map(v => v.message).join("\n\n")}
<p align="right">
  ${dangerSignaturePostfix(results, commitID)}
</p>
`
}

export function inlineTemplate(dangerID: string, results: DangerResults, file: string, line: number): string {
  const printViolation = (defaultEmoji: string) => (violation: Violation) => {
    const emojiString = `:${defaultEmoji}:`;
    return `- ${violation.icon || emojiString} ${violation.message}`
  }

  return `
<!--
${buildSummaryMessage(dangerID, results)}
${fileLineToString(file, line)}
-->
${results.fails.map(printViolation("no_entry_sign")).join("\n")}
${results.warnings.map(printViolation("warning")).join("\n")}
${results.messages.map(printViolation("book")).join("\n")}
${results.markdowns.map(v => v.message).join("\n\n")}
  `
}
