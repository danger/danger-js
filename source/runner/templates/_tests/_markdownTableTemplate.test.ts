import { template as markdownTableTemplate } from "../../templates/markdownTableTemplate"

describe("generating markdown tables", () => {
  it("build a markdown table with the content given", () => {
    const headers = ["", "Warnings"]
    const rows = [
      ["⚠️", "This is a very unimportant warning."],
      ["⚠️", "But, this warning is pretty important."],
    ]

    const table = markdownTableTemplate(headers, rows)

    expect(table).toMatchSnapshot()
  })
})
