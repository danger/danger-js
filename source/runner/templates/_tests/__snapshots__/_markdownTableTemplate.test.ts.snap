// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generating markdown tables build a markdown table with the content given 1`] = `
"|  | Warnings |
| --- | --- |
| ⚠️ | This is a very unimportant warning. |
| ⚠️ | But, this warning is pretty important. |"
`;
